sql: SELECT INTERVAL '+' produces CLI panic

Found with an experiment I'm doing with [go fuzz](https://github.com/dvyukov/go-fuzz).

```
root@:26257/> select interval '+';
E170730 00:08:07.727755 1 util/log/crash_reporting.go:82  a panic has occurred!
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 1 [running]:
github.com/cockroachdb/cockroach/pkg/util/log.RecoverAndReportPanic(0x9562858, 0xc4200783f0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:59 +0x95
panic(0x5486ae0, 0x656b650)
	/usr/local/Cellar/go/1.8/libexec/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/parser.sqlStdToDuration(0xc4201bad80, 0x1, 0x5731e58, 0xffffffffffffffff, 0x60e0840, 0xc42074c0c0, 0xc420696e60)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/interval.go:205 +0x1136
github.com/cockroachdb/cockroach/pkg/sql/parser.parseDInterval(0xc4201bad80, 0x1, 0x6, 0xc420696f10, 0xc420696f08, 0xc420204000)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/datum.go:1596 +0x1ef
github.com/cockroachdb/cockroach/pkg/sql/parser.ParseDInterval(0xc4201bad80, 0x1, 0xc42021e5e0, 0x3b, 0x161)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/datum.go:1513 +0x3e
github.com/cockroachdb/cockroach/pkg/sql/parser.(*sqlParserImpl).Parse(0xc42021e5e0, 0x60e5300, 0xc42021e500, 0x0)
	sql.y:5262 +0x53af9
github.com/cockroachdb/cockroach/pkg/sql/parser.(*Parser).Parse(0xc42021e500, 0xc42024e1e0, 0x14, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/parse.go:62 +0x6d
github.com/cockroachdb/cockroach/pkg/sql/parser.Parse(0xc42024e1e0, 0x14, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/parse.go:127 +0x4f
github.com/cockroachdb/cockroach/pkg/cli.(*cliState).doCheckStatement(0xc42063a900, 0x2, 0x4, 0xb, 0xb)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/cli/sql.go:774 +0x56
github.com/cockroachdb/cockroach/pkg/cli.runInteractive(0xc4205b0800, 0xc4204a1f00, 0x0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/cli/sql.go:941 +0x70a
github.com/cockroachdb/cockroach/pkg/cli.runTerm(0x6584320, 0xc4201f2440, 0x0, 0x1, 0x0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/cli/sql.go:996 +0x162
github.com/cockroachdb/cockroach/pkg/cli.MaybeDecorateGRPCError.func1(0x6584320, 0xc4201f2440, 0x0, 0x1, 0x0, 0x0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/cli/error.go:35 +0x5a
github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).execute(0x6584320, 0xc4201f2400, 0x1, 0x1, 0x6584320, 0xc4201f2400)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:644 +0x3f8
github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x65808a0, 0x402cf0e, 0xc42007e3b0, 0xc42069bf10)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:734 +0x339
github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).Execute(0x65808a0, 0xc42069bf00, 0xc42007e380)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:693 +0x2b
github.com/cockroachdb/cockroach/pkg/cli.Run(0xc42007c0a0, 0x2, 0x2, 0xc4200783f0, 0xc4200783f0)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:136 +0x6d
github.com/cockroachdb/cockroach/pkg/cli.Main()
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:51 +0x17f
main.main()
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/main.go:28 +0x20
```

--- a/pkg/sql/parser/interval.go
+++ b/pkg/sql/parser/interval.go
@@ -200,7 +200,10 @@ func sqlStdToDuration(s string) (duration.Duration, error) {
 				part = part[1:]
 			}
 		}
-		if part[0] == '-' {
+		if len(part) == 0 {
+			return d, errors.Errorf(errInvalidSQLDuration, s)
+		}
+		if part[0] == '-' || part[0] == '+' {
 			return d, errors.Errorf(errInvalidSQLDuration, s)
 		}
 

--- a/pkg/sql/parser/interval_test.go
+++ b/pkg/sql/parser/interval_test.go
@@ -17,7 +17,7 @@ package parser
 
 import "testing"
 
-func TestSQLIntervalSyntax(t *testing.T) {
+func TestValidSQLIntervalSyntax(t *testing.T) {
 	testData := []struct {
 		input  string
 		output string
@@ -83,6 +83,63 @@ func TestSQLIntervalSyntax(t *testing.T) {
 		if err != nil {
 			t.Errorf("%d: %q: %v", i, test.input, err)
 			continue
+		}
+		s := dur.String()
+		if s != test.output {
+			t.Errorf(`%d: %q: got "%s", expected "%s"`, i, test.input, s, test.output)
+			continue
+		}
+
+		dur2, err := parseDuration(s)
+		if err != nil {
+			t.Errorf(`%d: %q: repr "%s" is not parsable: %v`, i, test.input, s, err)
+			continue
+		}
+		s2 := dur2.String()
+		if s2 != s {
+			t.Errorf(`%d: %q: repr "%s" does not round-trip, got "%s" instead`,
+				i, test.input, s, s2)
+		}
+
+		// Test that a Datum recognizes the format.
+		di, err := parseDInterval(test.input, second)
+		if err != nil {
+			t.Errorf(`%d: %q: unrecognized as datum: %v`, i, test.input, err)
+			continue
+		}
+		s3 := di.Duration.String()
+		if s3 != test.output {
+			t.Errorf(`%d: %q: as datum, got "%s", expected "%s"`, i, test.input, s3, test.output)
+		}
+	}
+}
+
+func TestInvalidSQLIntervalSyntax(t *testing.T) {
+	testData := []struct {
+		input  string
+		output string
+		error  string
+	}{
+		{`+`, ``, `invalid input syntax for type interval +`},
+		{`++`, ``, `invalid input syntax for type interval ++`},
+		{`--`, ``, `invalid input syntax for type interval --`},
+	}
+	for i, test := range testData {
+		dur, err := sqlStdToDuration(test.input)
+		if err != nil {
+			if test.error != "" {
+				if err.Error() != test.error {
+					t.Errorf(`%d: %q: got error "%v", expected "%s"`, i, test.input, err, test.error)
+				}
+			} else {
+				t.Errorf("%d: %q: %v", i, test.input, err)
+			}
+			continue
+		} else {
+			if test.error != "" {
+				t.Errorf(`%d: %q: expected error "%q"`, i, test.input, test.error)
+				continue
+			}
 		}
 		s := dur.String()
 		if s != test.output {

